File: MultiplicativeNoiseLQR.m
% dynamic programming for linear systems with multiplicative noise

% problem data
n = 4;
m = 2;

% nominal system matrices (these are Abar and Bbar in the slides)
A = randn(n,n);
% scale down dynamics matrix to ensure mean-square stabilizability
A = 0.8*A/max(abs(eig(A))); 
B = randn(n,m);

% 
% friction = 10;  % nominal friction value
% A = [0 1; 
%      friction 1];
% 
% A1 = [0 0; 1 0];
% alpha1 = 1;  % variance of friction value

% multiplicative noise variances
SigmaA = randn(n^2, n^2); SigmaA = SigmaA*SigmaA'; 
SigmaA = 0.1*SigmaA/max(eig(SigmaA)); % scale down variances to ensure mean-square stabilizability
[Va,Ea] = eig(SigmaA);
a = diag(Ea); 
Aa = zeros(n,n,n^2);
for i=1:n^2
    Aa(:,:,i) = reshape(Va(:,i),n,n);
end

SigmaB = randn(n*m, n*m); SigmaB = SigmaB*SigmaB'; 
SigmaB = 0.1*SigmaB/max(eig(SigmaB)); % scale down variances to ensure mean-square stabilizability
[Vb,Eb] = eig(SigmaB);
b = diag(Eb); 
Bb = zeros(n,m,n*m);
for i=1:n*m
    Bb(:,:,i) = reshape(Vb(:,i),n,m);
end

Q = eye(n);
QT = eye(n);
R = eye(m);

T = 30;

%% finite horizon dynamic programming algorithm
P = zeros(n,n,T+1);
K = zeros(m,n,T);

P(:,:,T+1) = QT;

for i=T:-1:1
    P(:,:,i) = Q + MultSum(P(:,:,i+1), a, Aa) + A'*(P(:,:,i+1) - P(:,:,i+1)*B*inv(B'*P(:,:,i+1)*B + R)*B'*P(:,:,i+1))*A;
    K(:,:,i) = -inv(R + B'*P(:,:,i+1)*B + MultSum(P(:,:,i+1), b, Bb))*B'*P(:,:,i+1)*A;
end

%% infinite horizon
Pinf = zeros(n,n); % initialize
Pplus = randn(n,n); % data structure for optimal cost udpate

% value iteration recursion
t = 0;
while norm(Pinf - Pplus, 'fro') > 1e-6
    t = t + 1;
    Pinf = Pplus;
    Pplus = Q + A'*Pinf*A + MultSum(Pinf, a, Aa) - A'*Pinf*B*inv(R + B'*Pinf*B + MultSum(Pinf, b, Bb))*B'*Pinf*A;
end

Kinf = -inv(R + B'*Pinf*B + MultSum(Pinf, b, Bb))*B'*Pinf*A;

%% plot some stuff

plot(0:T-1,reshape(K(1,1,:),T,1), 0:T-1, reshape(K(1,2,:),T,1), 0:T-1, reshape(K(1,3,:),T,1), 0:T-1, reshape(K(1,4,:),T,1),'LineWidth',2)
xlabel('time');
ylabel('feedback gains');
legend('(K_t)_{11}', '(K_t)_{12}', '(K_t)_{13}', '(K_t)_{14}')
set(gca, 'fontsize', 20);


%% utility function for multiplicative noise variance terms
function Z = MultSum(X, z, Zz)
%%% 
    Z = zeros(size(Zz,2));
    for i=1:length(z)
        Z = Z + z(i)*Zz(:,:,i)'*X*Zz(:,:,i);
    end  
end

File: MultiplicativeNoiseLQR.py
import numpy as np
import matplotlib.pyplot as plt

# Utility function for multiplicative noise variance terms
def MultSum(X, z, Zz):
    Z = np.zeros((Zz.shape[1], Zz.shape[1]))
    for i in range(len(z)):
        Z = Z + z[i] * Zz[:, :, i].T @ X @ Zz[:, :, i]

    return Z


# Problem data
n = 4
m = 2

# Nominal system matrices
A = np.random.randn(n, n)
A = 0.8 * A / np.max(np.abs(np.linalg.eigvals(A)))

B = np.random.randn(n, m)

# Multiplicative noise variances for A and B
SigmaA = np.random.randn(n ** 2, n ** 2)
SigmaA = SigmaA @ SigmaA.T
SigmaA = 0.1 * SigmaA / np.max(np.linalg.eigvals(SigmaA))

a, Va = np.linalg.eig(SigmaA)
Aa = np.zeros((n, n, n ** 2))
for i in range(n ** 2):
    Aa[:, :, i] = Va[:, i].reshape(n, n)

SigmaB = np.random.randn(n * m, n * m)
SigmaB = SigmaB @ SigmaB.T
SigmaB = 0.1 * SigmaB / np.max(np.linalg.eigvals(SigmaB))

b, Vb = np.linalg.eig(SigmaB)
Bb = np.zeros((n, m, n * m))
for i in range(n * m):
    Bb[:, :, i] = Vb[:, i].reshape(n, m)

Q = np.eye(n)
QT = np.eye(n)
R = np.eye(m)

T = 30

# Finite horizon dynamic programming algorithm
P = np.zeros((n, n, T + 1))
K = np.zeros((m, n, T))

P[:, :, T] = QT

for i in range(T - 1, -1, -1):
    P[:, :, i] = Q + MultSum(P[:, :, i + 1], a, Aa) + A.T @ (P[:, :, i + 1] - P[:, :, i + 1] @ B @ np.linalg.inv(
                      B.T @ P[:, :, i + 1] @ B + R) @ B.T @ P[:, :, i + 1]) @ A

    K[:, :, i] = -np.linalg.inv(
        R + B.T @ P[:, :, i + 1] @ B + MultSum(P[:, :, i + 1], b, Bb)) @ B.T @ P[:, :, i + 1] @ A

# Infinite horizon
Pinf = np.zeros((n, n))
Pplus = np.random.randn(n, n)

# Value iteration recursion
t = 0
while np.linalg.norm(Pinf - Pplus, 'fro') > 1e-6:
    t += 1
    Pinf = Pplus
    Pplus = Q + A.T @ Pinf @ A + MultSum(Pinf, a, Aa) - A.T @ Pinf @ B @ np.linalg.inv(
        R + B.T @ Pinf @ B + MultSum(Pinf, b, Bb)) @ B.T @ Pinf @ A

Kinf = -np.linalg.inv(
    R + B.T @ Pinf @ B + MultSum(Pinf, b, Bb)) @ B.T @ Pinf @ A

# Plot feedback gains
time_steps = np.arange(T)
for j in range(m):
    plt.plot(time_steps, K[j, 0, :], label=f'(K_t)_{j + 1}1', linewidth=2)

plt.xlabel('Time')
plt.ylabel('Feedback Gains')
plt.legend()
plt.grid(True)
plt.show()